feat: implement timeline2 component - INNO-1546

haringsrob opened the pull request:
# PR description

Please drop a few lines about the PR: what it does, how to test it, etc.

## QA Checklist

In order to ensure a safe and quick review, please check that your PR follow those guidelines:

- [x] I have put the vanilla component as `devDependencies`
- [x] I have put the specs package as `devDependencies`
- [x] I have added the components directly used in the twig file (with `include` or `embed`) as `dependencies`
- [x] My component is listed in `@ecl-twig/ec-components`'s `dependencies`
- [x] My variables naming follow the guidelines (snake case for twig)
- [x] I have provided tests
- [x] I have provided documentation (for the "notes" tab)
- [x] If my local `yarn.lock` contains changes, I have committed it
- [x] I have given my PR the proper label (`pr: review needed` to indicate that I'm done and now waiting for a review ,`pr: wip` to indicate that I'm actively working on it ...)

yhuard commented on src/ec/packages/ec-component-timeline/timeline.html.twig:
```diff
@@ -0,0 +1,106 @@
+      {# Display the show more button if required.#}
+      {% if hide_amount > 0 and loop.revindex == 2 %}
+        <li class="ecl-timeline2__item ecl-timeline2__item--toggle">
+        <button data-ecl-timeline-button="true"
```
could you reuse the existing `@ecl-twig/ec-component-button` here? :)

haringsrob replied:
Done.

yhuard commented on src/ec/packages/ec-component-timeline/timeline.html.twig:
```diff
@@ -0,0 +1,106 @@
+{% spaceless %}
+
+{#
+  Parameters:
```
the parameters described here don't match the actual parameters of the templates

haringsrob replied:
Fixed

yhuard commented on src/ec/packages/ec-component-timeline/timeline.html.twig:
```diff
@@ -0,0 +1,106 @@
+        <div class="ecl-timeline2__title">{{ _item.title }}</div>
+        <div class="ecl-timeline2__content">
+          {%- set _content = _item.content|default('') %}
+          {%- block content _content -%}
```
I'm not sure this block is usable. How would a parent template embed the timeline and use the `content` block?

haringsrob replied:
Changed this to {{_content}}

haringsrob commented:
Hi @yhuard,

In addition to your comments some other fixes were made as well.

- Leftover "accordion" in the docs.
- A leftover <li> in the twig file.

The button is now using the component and I added the extra icon classes parameter.

I removed the button attribute from the component paremeters, not sure where I got that from..

Other remarks were due to copy pasting. Adjusted those.

haringsrob commented on src/ec/packages/ec-component-timeline/demo/data.js:
```diff
@@ -0,0 +1,8 @@
+import specData from '@ecl/ec-specs-timeline/demo/data';
+
+const adapter = initialData => {
+  return JSON.parse(JSON.stringify(initialData));
```
No it is not.

haringsrob replied:
I did leave the file but just return specData. I could remove the file as well and do it in the story itself?

yhuard commented on src/ec/packages/ec-component-timeline/package.json:
```diff
@@ -0,0 +1,32 @@
+  "description": "ECL Twig - EC Timeline",
+  "dependencies": {
+    "@ecl-twig/ec-component-button": "2.3.2",
+    "@ecl-twig/ec-component-icon": "2.3.2"
```
Now you don't need `"@ecl-twig/ec-component-icon": "2.3.2"` anymore since `@ecl-twig/ec-component-button` takes care of adding the icon

yhuard commented on src/ec/packages/ec-component-timeline/timeline.html.twig:
```diff
@@ -0,0 +1,96 @@
+              name: 'corner-arrow',
+              size: 'xs',
+            },
+            icon_extra_classes: 'ecl-icon--rotate-180',
```
You could use the [`transform`](https://github.com/ec-europa/ecl-twig/blob/v2.7-dev/src/ec/packages/ec-component-icon/icon.html.twig#L11) parameter of the icon component instead of manually adding the class 😉

haringsrob commented:
I removed the icon extra classes and instead now use the transform property.

yhuard commented on src/ec/packages/ec-component-timeline/docs/timeline.md:
```diff
@@ -0,0 +1,48 @@
+
+### Parameters
+
+- "toggleCollapsed": (string) (default: 'Show %d more items')
```
we follow the snake case convention when it comes to naming twig variables 😉

yhuard commented on src/ec/packages/ec-component-timeline/docs/timeline.md:
```diff
@@ -0,0 +1,48 @@
+
+- "toggleCollapsed": (string) (default: 'Show %d more items')
+- "toggleExpanded": (string) (default: 'Show %d items')
+- "hide" (array): (default: undefined)
```
```suggestion
- "hide" (object): (default: undefined)
```

yhuard commented on src/ec/packages/ec-component-timeline/docs/timeline.md:
```diff
@@ -0,0 +1,48 @@
+  hide: {from: 7, to: -2}, 
+  items: [ 
+    { 
+      id: '0', 
```
`id: '0'` is quite bad, can you change it here? (we will update the data from the specs later)

yhuard commented on src/ec/packages/ec-component-timeline/docs/timeline.md:
```diff
@@ -0,0 +1,48 @@
+- "toggleExpanded": (string) (default: 'Show %d items')
+- "hide" (array): (default: undefined)
+  - "from": (integer) (default: undefined) Positive integer
+  - "to": (integer) (default: undefined) Negative integer
```
didn't you want to change the description here?
```suggestion
  - "to": (integer) (default: undefined) Item index after which to resume displaying timeline items
```

yhuard commented on src/ec/packages/ec-component-timeline/timeline.html.twig:
```diff
@@ -0,0 +1,120 @@
+{% set _icon_path = icon_path|default('') %}
+{% set _toggle_collapsed = toggleCollapsed|default('Show %d more items') %}
+{% set _toggle_expanded = toggleExpanded|default('Hide %d items') %}
+{% set _hide_from = hide.from|default(items|length) %}
```
If `hide` is not defined, this could throw an error (unless `|default` also catches exceptions...).  Better safe than sorry :)

haringsrob replied:
Researched this and there is no need for this.

{{ hide.from|default('') }} works both here and in drupal without ever defining hide.

yhuard commented on src/ec/packages/ec-component-timeline/timeline.html.twig:
```diff
@@ -0,0 +1,120 @@
+{% set _hide_from = hide.from|default(items|length) %}
+{% set _hide_to = hide.to|default(items|length) %}
+
+{% set display_amount = items|length %}
```
all internal variables should be prepended with an underscore (sorry, it's an undocumented convention)

yhuard commented on src/ec/packages/ec-component-timeline/timeline.html.twig:
```diff
@@ -0,0 +1,120 @@
+      {% if hide_amount > 0 and loop.revindex == bottom_amount %}
+        <li class="ecl-timeline2__item ecl-timeline2__item--toggle">
+          {% include '../ec-component-button/button.html.twig' with {
+            label: toggleCollapsed|replace({'%d': countbase - display_amount}),
```
The component should be as dumb as possible, meaning it should not replace the labels. The labels should be provided with the right text.

yhuard commented on src/ec/packages/ec-component-timeline/timeline.html.twig:
```diff
@@ -0,0 +1,120 @@
+{% set bottom_amount = 0 %}
+
+{% if _hide_from > 0 %}
+  {% set bottom_amount = _hide_to|abs %}
```
what happens when `hide.to` is positive?

let's say I have 15 items, I want to hide from the 3rd to the 13th, I provide `hide: { from 3, to 13 }`
=> `bottom_amount = 13`
=> `countbase = 15 - 13 = 2`
=> `display_amount` = 3`
=> `hide_amount`= 2 - 3 = -1`
and nothing is hidden

the logic could be simplified

```
let first_item_to_hide, last_item_to_hide
if hide.from and hide.to are defined
  if hide.from < 0
    first_item_to_hide = items|length + hide.from
  else 
    first_item_to_hide = hide.from
  endif
  if hide.to < 0
    last_item_to_hide = items|length + hide.to
  else 
    last_item_to_hide = hide.to
  endif
endif

# now we have the right indices, let's make sure there's something to hide
bool hidden_items = first_item_to_hide < last_item_to_hide

# let's iterate
for item in items
  let item_css_class 
  if hidden_items == true
    if loop.index == last_item_to_hide + 1
      # print button
    else if first_item_to_hide < loop.index and loop.index < last_item_to_hide
      item_css_class = ecl-timeline2__item--collapsed
    endif
  endif
  # print item with item_css_class
endfor
```

I didn't care about the exact indices, but the overall logic could be simplified like this couldn't it?

haringsrob replied:
I have updated the calculation to take care of both scenarios.

yhuard commented on src/ec/packages/ec-component-timeline/timeline.test.js:
```diff
@@ -0,0 +1,66 @@
+    const optionsWithExtraClasses = merge(demoData, {
+      hide: {
+        from: 10,
+        to: 0,
```
would be nice to have a test scenario is a positive `hide.to` or with a negative `hide.from` :)

yhuard commented on src/ec/packages/ec-component-timeline/timeline.story.js:
```diff
@@ -0,0 +1,53 @@
+      const fullDemoData = {
+        ...demoData,
+        icon_path: iconPath,
+        hide: {
```
`hide` could (should?) come from `demoData`, not from the story
it might also be interesting to add a knob on the `hide.from` and `hide.to` values 😉

haringsrob commented:
Hi @yhuard I have taken care of the comments.

yhuard commented on src/ec/packages/ec-component-timeline/demo/data.js:
```diff
@@ -0,0 +1,26 @@
+  }
+
+  return {
+    toggle_collapsed: 'Show %d more items'.replace('%d', hiddenCount),
```
```suggestion
    toggle_collapsed: `Show ${hiddenCount} more items`,
```

yhuard commented on src/ec/packages/ec-component-timeline/demo/data.js:
```diff
@@ -0,0 +1,26 @@
+
+  return {
+    toggle_collapsed: 'Show %d more items'.replace('%d', hiddenCount),
+    toggle_expanded: 'Hide %d items'.replace('%d', hiddenCount),
```
```suggestion
    toggle_expanded: `Hide ${hiddenCount} items`,
```

yhuard commented on src/ec/packages/ec-component-timeline/demo/data.js:
```diff
@@ -0,0 +1,26 @@
+  if (to > 0) {
+    hiddenCount = to - from;
+  } else {
+    hiddenCount = specData.items.length - (from + Math.abs(to));
```
or simpler:
```suggestion
    hiddenCount = initialData.items.length + to - from;
```

yhuard commented on src/ec/packages/ec-component-timeline/timeline.story.js:
```diff
@@ -0,0 +1,65 @@
+      if (to > 0) {
+        hiddenCount = to - from;
+      } else {
+        hiddenCount = demoData.items.length - (from + Math.abs(to));
```
same remark as earlier, it could be simplified a bit :)

yhuard commented on src/ec/packages/ec-component-timeline/timeline.story.js:
```diff
@@ -0,0 +1,65 @@
+  .add(
+    'default',
+    () => {
+      const from = number('From', 10);
```
what about using `demoData.hide.from` as a default instead of 10? (same for `to`) 
thus the changes made to `demo/data.js` would be reflected here as well

yhuard commented on src/ec/packages/ec-component-timeline/timeline.story.js:
```diff
@@ -0,0 +1,65 @@
+
+      const fullDemoData = {
+        ...demoData,
+        toggle_collapsed: 'Show %d more items'.replace('%d', hiddenCount),
```
same remark as earlier, template literals are preferable (more JS-y)

yhuard commented on src/ec/packages/ec-component-timeline/timeline.html.twig:
```diff
@@ -0,0 +1,101 @@
+    data-ecl-timeline="true">
+  {% if _items is not empty %}
+    {% for _item in _items %}
+      {# Deside on per item if it should display.#}
```
```suggestion
      {# Decide on per item if it should display. #}
```

yhuard commented on src/ec/packages/ec-component-timeline/timeline.html.twig:
```diff
@@ -0,0 +1,101 @@
+{% set _hide_amount = 0 %}
+{% set _bottom_amount = 0 %}
+
+{% if _hide_from > 0 %}
```
Which means a negative `from` index is not supported yet (this feature could be added later if needed).

This is not a blocker ;)
